canister/src/api/get_block_headers.rs

```diff
@@ -1,17 +1,130 @@
+                    end_height,
+                });
+            }
+
+            if end_height > chain_height {
+                return Err(GetBlockHeadersError::EndHeightDoesNotExist {
+                    requested: end_height,
+                    chain_height,
+                });
+            }
+            // If `end_height` is provided then it should be the
+            // end of effective height range.
+            (request.start_height, end_height)
+        } else {
+            // If `end_height`` is not provided then the end of effective
+            // range should be the last block of the chain.
+            (request.start_height, chain_height)
+        };
+
+    // Bound the length of block headers vec.
+    effective_end_height = std::cmp::min(
+        effective_end_height,
+        effective_start_height + MAX_BLOCK_HEADERS_PER_RESPONSE - 1,
+    );
+
+    Ok((effective_start_height, effective_end_height))
+}
+
+fn get_block_headers_internal(
+    request: &GetBlockHeadersRequest,
+) -> Result<(GetBlockHeadersResponse, Stats), GetBlockHeadersError> {
+    let (start_height, end_height) = verify_and_return_effective_range(request)?;
+
+    let mut stats: Stats = Stats::default();
+
+    // Build block headers vec.
+    let ins_start = performance_counter();
+
+    // Add requested block headers located in stable_blocks.
+    let mut vec_headers: Vec<Vec<u8>> = with_state(|s| {
+        let block_heights = &s.stable_block_headers.block_heights;
+        let block_headers = &s.stable_block_headers.block_headers;
+        block_heights
+            .range(start_height..=end_height)
+            .map(|(_, block_hash)| block_headers.get(&block_hash).unwrap().into())
+            .collect()
+    });
+
+    // The last stable block is located in `unstable_blocks`, the height of the
+    // first block in `unstable_blocks` is equal to `stable_height`.
+    let height_of_first_block_in_unstable_blocks = with_state(|s| s.stable_height());
+
+    // Add requested block headers located in unstable_blocks.
+    if end_height >= height_of_first_block_in_unstable_blocks {
+        let start_range_in_unstable_blocks =
+            if start_height < height_of_first_block_in_unstable_blocks {
+                0
+            } else {
+                start_height - height_of_first_block_in_unstable_blocks
+            };
+
+        let end_range_in_unstable_blocks = end_height - height_of_first_block_in_unstable_blocks;
+
+        with_state(|s| {
+            let unstable_blocks = s.get_unstable_blocks_in_main_chain().into_chain();
+
+            for i in start_range_in_unstable_blocks..=end_range_in_unstable_blocks {
+                let mut header_blob = vec![];
+                unstable_blocks[i as usize]
+                    .header()
+                    .consensus_encode(&mut header_blob)
+                    .unwrap();
+                vec_headers.push(header_blob);
+            }
+        });
+    }
```

Comment:
Same here. Try to encapsulate as much logic in the respective components as possible. In this case, this logic can be added into the `UnstableBlocks` component. Perhaps something like this:

```
    pub fn get_block_headers_in_range(
        &self,
        stable_height: Height,
        heights: std::ops::RangeInclusive<Height>,
    ) -> impl Iterator<Item = &BlockHeader> {

```

Reply:
Done.